### PR TITLE
Add Method to Convert QgsRasterLayer to NumPy Array

### PR DESCRIPTION
**Pull request by merydian**
## Description

This pull request, as part of qgis/QGIS-Enhancement-Proposals#227, introduces a new method to PyQGIS that enables conversion of `QgsRasterLayer` into a NumPy array. This addition allows users to efficiently handle and analyze raster data using NumPy. This enhancement improves integration with NumPy while keeping it an optional dependency.

### CODE REVIEW

**Comment by nyalldawson** [comment]
@merydian can you update this branch based on the raster block work?

**Comment by nyalldawson** [comment]
@merydian can you rebase against master? There's a merge conflict here now

File: python/PyQt6/core/__init__.py.in [hunk 1]
```diff
@@ -515,12 +515,30 @@ try:
 
    QgsRasterBlock.as_numpy = _raster_block_as_numpy
 
+   def _raster_layer_as_numpy(self, use_masking=True, bands: _typing.Optional[_typing.List[int]] = None):
```
Comment (nyalldawson) [inline python/PyQt6/core/__init__.py.in:518]: Is it possible to add the type hint for the return type here?

File: python/PyQt6/core/__init__.py.in [hunk 2]
```diff
@@ -530,3 +548,12 @@ If `use_masking` is `True` then the returned array will be a numpy masked array,
 
 .. versionadded:: 3.40
 """
+QgsRasterLayer.as_numpy.__doc__ = """
+Returns the layer data as a numpy array.
+
+If `use_masking` is `True` then the returned arrays will be numpy masked arrays, masking the raster block's nodata values.
+
```
Comment (nyalldawson) [inline python/PyQt6/core/__init__.py.in:555]: Can you add documentation for the bands argument too?

File: python/PyQt6/core/__init__.py.in (hunk 1 above)
Comment (merydian) [reply python/PyQt6/core/__init__.py.in:518]: I guess it should be ` -> _numpy.ndarray`?

Comment (nyalldawson) [reply python/PyQt6/core/__init__.py.in:518]: Actually `_typing.List[_typing.Union[_numpy.ndarray, _numpy.ma.MaskedArray]]`

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
Perfect, thank you!

**Pull request merged**